File: Day173_2025/Q173_GFG.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

class Q173_GFG {
    public ArrayList<Integer> largestSubset(int[] arr) {
        int n = arr.length;
        Arrays.sort(arr); // ascending order
        int[] dp = new int[n];
        int[] parent = new int[n];
        Arrays.fill(dp, 1);
        Arrays.fill(parent, -1);

        int maxLen = 0, maxIdx = -1;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (arr[i] % arr[j] == 0 && dp[i] < dp[j] + 1) {
                    dp[i] = dp[j] + 1;
                    parent[i] = j;
                }
            }
            if (dp[i] > maxLen) {
                maxLen = dp[i];
                maxIdx = i;
            }
        }

        ArrayList<Integer> result = new ArrayList<>();
        while (maxIdx != -1) {
            result.add(arr[maxIdx]);
            maxIdx = parent[maxIdx];
        }

        // optional: reverse the list to maintain increasing order
        Collections.reverse(result);
        return result;
    }
}

File: Day173_2025/Q173_GFG.py
class Solution:
    def largestSubset(self, arr):
        n = len(arr)
        arr.sort()
        dp = [1] * n
        parent = [-1] * n
        max_len = 0
        max_idx = -1

        for i in range(n):
            for j in range(i):
                if arr[i] % arr[j] == 0 and dp[i] < dp[j] + 1:
                    dp[i] = dp[j] + 1
                    parent[i] = j
            if dp[i] > max_len:
                max_len = dp[i]
                max_idx = i

        result = []
        while max_idx != -1:
            result.append(arr[max_idx])
            max_idx = parent[max_idx]
        result.reverse()
        return result
